pkg/provider/ecs/ecs.go

```diff
@@ -308,74 +282,71 @@ func (p *Provider) listInstances(ctx context.Context, client *awsClient) ([]ecsI
 		}
 
 		for key, task := range tasks {
-			containerInstance := ec2Instances[aws.StringValue(task.ContainerInstanceArn)]
+			containerInstance, hasContainerInstance := ec2Instances[aws.ToString(task.ContainerInstanceArn)]
 			taskDef := taskDefinitions[key]
 
 			for _, container := range task.Containers {
-				var containerDefinition *ecs.ContainerDefinition
+				var containerDefinition *ecstypes.ContainerDefinition
 				for _, def := range taskDef.ContainerDefinitions {
-					if aws.StringValue(container.Name) == aws.StringValue(def.Name) {
-						containerDefinition = def
+					if aws.ToString(container.Name) == aws.ToString(def.Name) {
+						containerDefinition = &def
 						break
 					}
 				}
 
 				if containerDefinition == nil {
-					logger.Debug().Msgf("Unable to find container definition for %s", aws.StringValue(container.Name))
+					logger.Debug().Msgf("Unable to find container definition for %s", aws.ToString(container.Name))
 					continue
 				}
 
 				var mach *machine
-				if aws.StringValue(taskDef.NetworkMode) == "awsvpc" && len(task.Attachments) != 0 {
+				if taskDef.NetworkMode == ecstypes.NetworkModeAwsvpc && len(task.Attachments) != 0 {
 					if len(container.NetworkInterfaces) == 0 {
-						logger.Error().Msgf("Skip container %s: no network interfaces", aws.StringValue(container.Name))
+						logger.Error().Msgf("Skip container %s: no network interfaces", aws.ToString(container.Name))
 						continue
 					}
 
 					var ports []portMapping
 					for _, mapping := range containerDefinition.PortMappings {
-						if mapping != nil {
-							protocol := "TCP"
-							if aws.StringValue(mapping.Protocol) == "udp" {
-								protocol = "UDP"
-							}
-
-							ports = append(ports, portMapping{
-								hostPort:      aws.Int64Value(mapping.HostPort),
-								containerPort: aws.Int64Value(mapping.ContainerPort),
-								protocol:      protocol,
-							})
+						protocol := ecstypes.TransportProtocolTcp
+						if mapping.Protocol == ecstypes.TransportProtocolUdp {
+							protocol = ecstypes.TransportProtocolUdp
 						}
+
+						ports = append(ports, portMapping{
+							hostPort:      aws.ToInt32(mapping.HostPort),
+							containerPort: aws.ToInt32(mapping.ContainerPort),
+							protocol:      protocol,
+						})
```

Comment:
AWS SDK V2 no longer uses pointer _literally_ everywhere. In this case, V1 and V2 has different types for `ContainerDefinition.PortMappings`:

- V1 is `[]*PortMapping`: https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.55.5/service/ecs#ContainerDefinition.PortMappings
- V2 is `[]PortMapping`: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ecs/types@v1.53.0#ContainerDefinition.PortMappings

So we don't need the `if mapping != nil` if statement here anymore.

The same applies to other areas in this file.